monai/transforms/intensity/array.py

```diff
@@ -1007,6 +1008,139 @@ def __call__(self, img: NdarrayOrTensor) -> NdarrayOrTensor:
+                if soft clipping and percentile is None, return None as the corresponding percentile in meta information.
+            return_percentiles: whether to return the calculated percentiles in tensor meta information,
+                if soft clipping and percentile is None, return None as the corresponding percentile in meta information.
+                defaults to False.
+            dtype: output data type, if None, same as input image. defaults to float32.
+        """
+        if lower is None and upper is None:
+            raise ValueError("lower or upper percentiles must be provided")
+        if lower is not None and (lower < 0.0 or lower > 100.0):
+            raise ValueError("Percentiles must be in the range [0, 100]")
+        if upper is not None and (upper < 0.0 or upper > 100.0):
+            raise ValueError("Percentiles must be in the range [0, 100]")
+        if upper is not None and lower is not None and upper < lower:
+            raise ValueError("upper must be greater than or equal to lower")
+        if sharpness_factor is not None and sharpness_factor <= 0:
+            raise ValueError("sharpness_factor must be greater than 0")
+
+        self.lower = lower
+        self.upper = upper
+        self.sharpness_factor = sharpness_factor
+        self.channel_wise = channel_wise
+        if return_percentiles:
+            self.percentiles = []
+        self.return_percentiles = return_percentiles
+        self.dtype = dtype
+
+    def _clip(self, img: NdarrayOrTensor) -> NdarrayOrTensor:
+        if self.sharpness_factor is not None:
+            lower_percentile = percentile(img, self.lower) if self.lower is not None else None
+            upper_percentile = percentile(img, self.upper) if self.upper is not None else None
+            img = soft_clip(img, self.sharpness_factor, lower_percentile, upper_percentile, self.dtype)
+        else:
+            lower_percentile = percentile(img, self.lower) if self.lower is not None else percentile(img, 0)
+            upper_percentile = percentile(img, self.upper) if self.upper is not None else percentile(img, 100)
+            img = clip(img, lower_percentile, upper_percentile)
+
+        if self.return_percentiles:
+            self.percentiles.append((lower_percentile, upper_percentile))
+        img = convert_to_tensor(img, track_meta=False)
+        return img
+
+    def __call__(self, img: NdarrayOrTensor) -> NdarrayOrTensor:
+        """
+        Apply the transform to `img`.
+        """
+        img = convert_to_tensor(img, track_meta=get_track_meta())
+        img_t = convert_to_tensor(img, track_meta=False)
+        if self.channel_wise:
+            img_t = torch.stack([self._clip(img=d) for d in img_t])  # type: ignore
+        else:
+            img_t = self._clip(img=img_t)
+
+        img = convert_to_dst_type(img_t, dst=img)[0]
+        if self.return_percentiles:
+            img.meta["percentiles"] = self.percentiles
```

Comment:
How about changing "percentiles" to "clipping_values" to make the naming clearer in meta tensor?

Reply:
Done! You're right, it's clearer that way.

What do you think of the structure and type of these clipping values? As a list of 2d-tuple containing either a NdarrayOrTensor or a None? That's the part I'm a bit dubious about. As far as I'm concerned it's better that way but it's also much less manipulable and perhaps not very clear for users.

Reply:
The data structure looks good to me. You can add a few words in the docstrings to explain `None`. And we can unify the data type to "float". Since these values are individual numbers, floats should support numpy arrays or torch tensors for potential use. What do you think?

Reply:
Okay, as it is currently done, this returns a list of tuples containing either a float or a None.
so the corresponding typing should be something like: ```List[Tuple[Optional[float], ...]]```

Reply:
I've added typing and instead of returning  0-dim tensor or array I chose to keep only the float value via the `.item()`. Does that seem better to you?

Reply:
Does it work for numpy arrays?

Reply:
Yes, `item()` can also be applied to 0-d numpy arrays but here with the `conversion_to_tensor` if it is a numpy at first it will be converted to a `torch.Tensor`

Reply:
I see.